layers/water-material.js

```diff
@@ -0,0 +1,346 @@
+          float linearEyeDepth = getViewZ(getDepth(screenUV));
+
+          if (vPos.y <= 0.01) { // ocean shader, for temporary. TODO: We should have attribute to indicate the water type(ocean, river, waterfall, etc.,)
+            if (!cameraInWater) {
+              //################################## compute waterColor ##################################
+              float depthScale = 15.;
+              float depthFalloff = 3.;
+              float sceneDepth = getDepthFade(fragmentLinearEyeDepth, linearEyeDepth, depthScale, depthFalloff);
+
+              // set green water color below player. 
+              vec3 viewIncidentDir = normalize(eye - vWorldPosition.xyz);
+              vec3 viewReflectDir = reflect(viewIncidentDir, vec3(0., 1.0, 0.));
+              float fresnelCoe = (dot(viewIncidentDir,viewReflectDir) + 1.) / 2.;
+              fresnelCoe = clamp(fresnelCoe, 0., 1.0);
+              float waterColorDepth = getDepthFade(fragmentLinearEyeDepth, linearEyeDepth, 20., 1.);
+              float colorLerp = mix(fresnelCoe, 1. - waterColorDepth, waterColorDepth);
+              colorLerp = mix(colorLerp, 1. - waterColorDepth, saturate(distance(eye, vWorldPosition) / 150.));
+
+              // water color
+              vec4 cos_grad = cosine_gradient(saturate(1. - colorLerp), phases, amplitudes, frequencies, offsets);
+              cos_grad = clamp(cos_grad, vec4(0.), vec4(1.));
+              vec4 waterColor = vec4(cos_grad.rgb, 1. - sceneDepth);
+              
+              //################################## handle foam ##################################
+              float fadeoutDistance = 50.;
+              float fadeoutScale = 3.;
+              float fadeoutLerp = pow(saturate(distance(playerPos, vWorldPosition) / fadeoutDistance), fadeoutScale);
+          
+              // foam distortion
+              vec4 ds2 = texture2D( 
+                tDistortion, 
+                vec2(
+                  0.25 * vWorldPosition.x + uTime * 0.01,
+                  0.25 * vWorldPosition.z + uTime * 0.01
+                ) 
+              );
+              vec4 ds = texture2D( 
+                tDistortion, 
+                vec2(
+                  0.3 * vWorldPosition.x + uTime * 0.005,
+                  0.3 * vWorldPosition.z + uTime * 0.005
+                ) 
+              );
+              float foamDistortionDepth = getDepthFade(fragmentLinearEyeDepth, linearEyeDepth, 8., 2.);
+              float foamDistortionScale = 0.2;
+              float foamDistortionDegree = pow(clamp((1. - foamDistortionDepth), 0.2, 1.0), foamDistortionScale);
+              vec2 foamDistortion = vec2(ds2.r + ds.r, ds2.g + ds.g) * foamDistortionDegree;
+
+              // foam
+              float foamDepth = getDepthFade(fragmentLinearEyeDepth, linearEyeDepth, 8., 2.);
+              float foamDiff = saturate( fragmentLinearEyeDepth - linearEyeDepth );
+              float foamScale = -0.02;
```

Comment:
can we have these constant variables at the top of the file or possibly you can do something like this in the shader :
```
float FOAM_SCALE = ${WATER_FOAM_SCALE};
```
and get the constants from JS
( just a small note but I think this will help the code's readability )